README
README FOR THE Pod-Spelling DISTRIBUTION

This package contains a number of modules for spell-checking POD
using either Lingua::Ispell or Text::Aspell. One of those modules
must be installed on your system, with their binaries, unless you
plan to use the API to provide your own spell-checker.

INSTALLATION

	perl Makefile.PL
	make
	make test
	sudo make install
	


Changes
CHANGES

0.23 2015-06-17
	Added support for `=for stopwords` and `=for spelling` 
	and `=for allowed_words`.

0.22 2015-06-*
	Various bug fixes and a bit of a tidy-up

0.17 2013-04-04
	Added skip_paths_matching()
	Fixed index error
	Spellers split at dashes, underscores, and numbers, but try
	to allow that to pass.
	Crudely test for subs using UNIVERSAL::can
		
0.16 2013-02-26
	Updated Makefile.PL to accomodate Lingua::Ispell 0.07 API change
	(no rv for correctly-spelt words)
	RT 

0.15 2012-04-12
	Updated non-word characters removed from words prior to spellcheck
	
0.14 2012-04-11
	Allow *.yml files from more recent Perls than originally used in dev

0.13 ...

0.12 2011-03-11
	Try to catch bad installs of binary spellers
	
0.11 2011-03-10
	Ispell.pm exists but Ispell not found

0.5-0.10 2011-02
	Tests

0.4 2011-01-11 
	Removed Test::Most

0.3 2011-01-10
	Forgot the Makefile.PL and manifest tests didn't pick it up.
	
0.2 2011-01-10
	Just found Test::Spelling, so made this module API-compatible
	
0.1	2011-01-10
	Initial version
	


lib/Test/Pod/Spelling.pm
use strict;
use warnings;
use utf8;

package Test::Pod::Spelling;
our $VERSION = '0.4';

=encoding utf8

=head1 NAME

Test::Pod::Spelling - A Test library to spell-check POD files

=head1 SYNOPSIS

	use Test::Pod::Spelling;
	all_pod_files_spelling_ok();
	add_stopwords(qw( Goddard inine ));
	done_testing();

	use Test::Pod::Spelling (
		spelling => {
				allow_words => [qw[ 
					Goddard LICENCE inline behaviour spelt
				]]
			},
		);
	};
	pod_file_spelling_ok( 't/good.pod' );
	all_pod_files_spelling_ok();
	done_testing();
	
=head1 DESCRIPTION

This module exports two routines, described below, to test POD for spelling errors,
using either Lingua::Ispell and Text::Aspell. One of those modules
must be installed on your system, with their binaries, unless you
plan to use the API to provide your own spell-checker.

As illustrated in L</SYNOPSIS>, above, configuration options for C<Pod::Spelling> can
be passed when the module is used.

A list of words that can be allowed even if not in the dictionary (a stop list)
can be supplied to the spell-checking module when this module is used, via
L</add_stopwords>.
To help keep this list short, common POD that would upset the spell-checker
is skipped - see L<Pod::Spelling/TEXT NOT SPELL-CHECKED> for details.
No support is offered for inlining stoplists into Perl modules, which the
author feels undermines testing. Patches to provide such an option will be
applied, though, if required.

=cut

use base qw( 
	Test::Builder::Module 
	Exporter 
);

my $CLASS = __PACKAGE__;

=head1 DEPENDENCIES

L<Test::Pod|Test::Pod>,
L<Pod::Spelling|Pod::Spelling>,
L<Test::Builder|Test::Builder>.

=cut

use Test::Builder;
require Test::Pod;
require Pod::Spelling;
use Carp;

my $Test = Test::Builder->new;
$Test->{_skip_files} = {};

=head1 EXPORTS

	all_pod_files_spelling_ok() 
	pod_file_spelling_ok() 

=cut

sub import {
    my $self = shift;
    my @args = @_;
    my $spelling_args = {};
    
    # Get the spelling argument:
    for my $i (0..$#args){
    	if ($args[$i] and $args[$i] eq 'spelling'){
    		confess 'During import, the "spelling" argument must point to a HASH or Pod::Spelling object'
    			if $i==$#args
    			or not ref($args[$i+1]) 
    			or ref($args[$i+1]) !~ /^(HASH|Pod::Spelling.*)$/; 
    		# Use to init obj later
    		$spelling_args = $args[$i+1];
			# Remove from args that will be passed to plan()
    		@args = @args[
				0..$i-1,
				$i+2 .. $#args
			];
    	}
    	if ($args[$i] and $args[$i] eq 'skip'){
    		confess 'During import, the "skip" argument must point to an ARRAY'
    			if $i==$#args
    			or not ref($args[$i+1]) 
    			or ref($args[$i+1]) ne 'ARRAY'; 
    		$Test->{_skip_files} = { map {$_=>1} $args[$i+1] };
			# Remove from args that will be passed to plan()
    		@args = @args[
				0..$i-1,
				$i+2 .. $#args
			];
		}
    }
    
    $Test->{_speller} = Pod::Spelling->new( $spelling_args );
    
    my $caller = caller;

    for my $func ( qw(
		add_stopwords all_pod_files_spelling_ok pod_file_spelling_ok
		skip_paths_matching
	)) {
        no strict 'refs';
        *{$caller."::".$func} = \&$func;
    }

    $Test->exported_to($caller);
    $Test->plan(@args);
}

=head1 EXPORTED SUBROUTINES

=head2 C<skip_paths_matching($re1 [, $reN])>

Supply a list of one or more pre-compiled regular
expressions to skip any file paths they match.

=cut

sub skip_paths_matching {
	my @res = @_;
    local $Test::Builder::Level = $Test::Builder::Level + 1;
	my $rv = 1;
	# Make sure they are regex
	foreach my $i (@res){
		if (not ref $i or ref $i ne 'Regexp'){
			$rv = 0;
		}
		else {
			$Test->{_speller}->skip_paths_matching( $i );
		}
	}
	return $rv;
}

=head2 C<all_pod_files_spelling_ok( [@entries] )>

Exactly the same as L<Test::Pod/all_pod_files_ok( [@entries] )>
except that it calls L<Test::Pod/pod_file_ok( FILENAME[, TESTNAME ] )>
to check the spelling of POD files.

=cut

sub all_pod_files_spelling_ok {
	my @args = @_ ? @_ : Test::Pod::_starting_points();
    my @paths = map { -d $_ ? Test::Pod::all_pod_files($_) : $_ } @args;
    my @errors;
    
    PATH:
	foreach my $path (@paths){
		foreach my $re ($Test->{_speller}->skip_paths_matching){
			if ($path =~ $re){
				$Test->skip("path $path because of skip_paths_matching $re");
				next PATH;
			}
		}
        # local $Test::Builder::Level = $Test::Builder::Level + 1;
    	push @errors, pod_file_spelling_ok( $path );
    }

    return keys %{{
    	map {$_=>1} grep {defined} @errors
	}};
}

=head2 C<pod_file_spelling_ok( FILENAME[, TESTNAME ] )>

Exactly the same as L<Test::Pod/pod_file_ok( FILENAME[, TESTNAME ] )>
except that it checks the spelling of POD files.

=cut

sub pod_file_spelling_ok {
	my ($path, $name) = @_;
	return () if exists $Test->{_skip_files}->{$path};
	
	# All good POD has =head1 NAME\n\n$TITLE - $DESCRIPTION
	# so add that title to the dictionary. It may be a script name
	# without colons, so using the module name or path is not enough.
	open my $IN, $path or confess 'Could not open '.$path;
	read $IN, my $file, -s $IN;
	close $IN;
	my ($pod_name) = $file =~ /^=head1\s+NAME[\n\r\f]+\s*(\S+)\s*-\s*/m;
	undef $file;	
	
	if ($pod_name){
		my @words = $pod_name =~ s/:+/ /g;
		$Test->{_speller}->add_allow_words(
			$pod_name,
			split(/\s+/, @words)
		);
	}
	
	if (not $name){
		$name = 'POD spelling test for '. ( $pod_name || $path );
	}
	
	my @errors = $Test->{_speller}->check_file( $path );

	$Test->ok( not( scalar @errors), $name );

	if (@errors){
		foreach my $line ( 0 .. $#{ $Test->{_speller}->{errors} }){
			my $misspelt = $Test->{_speller}->{errors}->[$line];
			if ($misspelt and scalar @$misspelt){
				$Test->diag( sprintf
					'  Unknown word%s in %s (POD line %d): %s.',
					(scalar @$misspelt==1? '':'s'),
					$path, ($line+1), 
					join ', ', map("\"$_\"", @$misspelt)
				);
			}
		}
	}
	
	return @errors;
}

=head2 add_stopwords( @words )

Adds a list of stop-words to those already being used by the spell-checker.

=cut

sub add_stopwords {
	$Test->{_speller}->add_allow_words( @_ );
	return 1;
}

1;

__END__

=head1 TODO

Automatically skip the name of the author as described in 
F<Makefile.PL> or F<Build.PL> or similar.

=head1 AUTHOR AND COPYRIGHT

Copyright Lee Goddard (C) 2011. All Rights Reserved.

Made available under the same terms as Perl.



lib/Pod/Spelling.pm
use strict;
use warnings;
use utf8;

package Pod::Spelling;
our $VERSION = 0.6; # Catch undefined

use Pod::POM;
require Pod::POM::View::TextBasic;
use warnings::register;
use Carp;

sub new {
	my ($class, $args) = (
		(ref($_[0])? ref($_[0]) : shift ),
		(ref($_[0])? shift : {@_})
	);

	# Pod::POM->default_view( 'Pod::POM::View::TextBasic' )
	# 	or confess $Pod::POM::ERROR;

	my $self = bless {
		%$args,
		_parser => Pod::POM->new,
		_temp_stoplist => [],
	}, $class;
	
	$self->{skip_paths_matching} ||= [];

	$self->{view} ||= 'Pod::POM::View::TextBasic';

	# Allow a single word to be allowed:
	if ($self->{allow_words} and not ref $self->{allow_words}){
		$self->{allow_words} = [ $self->{allow_words} ];
	}

	$Pod::Wordlist::Wordlist = {};
	unless ($self->{not_pod_wordlist}){
		eval { 
			no warnings;
			require Pod::Wordlist;
		};
		warnings::warnif( $@ ) if $@;
	}
	
	if (ref $self and $self->{import_speller}){
		$self->import_speller( $self->{import_speller} );
	}

	# If no speller was specified and no callback provided,
	# try to find one of the defaults.	
	else {
		if (not $self->{spell_check_callback}){
			foreach my $mod (qw( Hunspell Ispell Aspell )){
				last if $self->import_speller( 'Pod::Spelling::'.$mod );
			}
		}
		$self = $self->_init;
	}

	Carp::confess 'Could not instantiate any spell checker. Do you have Hunspell, Ispell or Aspell installed with dictionaries?'
		if not $self->{spell_check_callback};

	return $self;
}

# Abstract method to be implemented by sub-classes:
# if AOK, return calling object, otherwise error string.
sub _init { return $_[0] }

sub import_speller {
	my ($self, $class) = @_;

	eval "require $class";	

	if ($@){
		warnings::warnif($@);
		$self->{spell_check_callback} = undef;
		return undef;
	}
	else {
		my $method = $class.'::_init';
		$self = $self->$method;
		$self->{spell_check_callback} = $class."::_spell_check_callback"
			if ref $self;
	}
	
	return ref $self;
}

# Method that accepts one or more lines of text, returns a list mispelt words.
sub _spell_check_callback {
	my $self = shift;
	warnings::warnif( 
		'No spell_check_callback registered: no spell checking is happening!'
	);
	# Return all words as errors
	return split /\s+/, join "\n", @_;	
}


sub _clean_text {
	my ($self, $text) = @_;
	return '' if not $text;
	
	$text =~ s/(\w+::)+\w+/ /gs;	# Remove references to Perl modules
	$text =~ s/\s+/ /gs;
	$text =~ s/[^\w-]+/ /gs;		# Remove punctuation
	$text =~ s/\d+//sg;
	$text =~ s/["'](\w+)["']/$1/sg;
	$text =~ s/\b-(\w+)/ $1/sg;
	$text =~ s/(\w+)-\b/$1 /sg;
	$text =~ s/\s+\W\s+/ /sg;
	$text =~ s/\bPerl\b/ /sg;
	
	foreach my $word ( @{$self->{allow_words}} ){
		next if not defined $word;
		$text =~ s/\b\Q$word\E\b//sig;
	}

	unless (exists $self->{no_pod_wordlist} or exists $self->{no_pod_wordlist}){
		no warnings 'once';
		foreach my $word (split /\s+/, $text){
			$word = '' if exists $Pod::Wordlist::Wordlist->{$word};
		}
	}

	# Allow words that are joined by underscores but
	# which were thought errors: easier than parsing
	# Perl, for now:
	$text =~ s/(\w+_\w+||_\w+||\w_)//sg;

	return $text;
}


# Returns all badly spelt from the file,
# and sets $self->{errors}->[ $line_number-1 ]->[ badly spelt words for this line ]
sub check_file {
	my ($self, $path) = @_;
    my ($packages, @rv);
    $self->{errors} = [];
	
	foreach my $re (@{ $self->{skip_paths_matching} }){
		return () if $path =~ $re;
	}

	# Crude test to allow package names:
	{
		# Hope it is not too large.
		open my $IN, $path or confess "$! - $path";
		read $IN, my $content, -s $IN;
		close $IN;

		my @packages = grep {length} $content =~ /^([}{;]+||)\s*package\s+([a-z](?:[\w]+||::)*?)\s*[;{]/img;

		$self->add_allow_words(
			@packages,				# Whole package names
			# Parts of package names
			map {
				split /::/, $_
			} @packages
		);
	}

	# To support '=for stopwords', we could create yet another parser, 
	# and reparse the document, but not sure why that would be better
	# than this:

		
    my $pom = $self->{_parser}->parse_file($path)
    	or confess $self->{_parser}->error();

	my $code = $self->{spell_check_callback};
	
	my $line = 0;
	foreach my $node ($pom->content){

		if ($node->type() =~ /^(begin|for)$/){
			my $allowed_line = $node->present( $self->{view} );
			my @stoplist = split /\s/, $allowed_line;
			$self->_add_temporary_stoplist( @stoplist );
			next;
		}

		my $text = $node->present( $self->{view} );
		$text =~ s/[\n\r\f]+/ /sg;

		$text = $self->_clean_text( $text );
		my @err = $self->$code( $text );
		
		ERR:
		foreach my $err (@err){
			foreach my $word (grep {length} split /\s+/, $text){
				if ($word =~ /([\d_-]\Q$err\E|\Q$err\E[\d_-])/sg){
					$err = undef;
					next ERR;
				}
			}
			# Check packages
			eval {
				require $path;
				foreach my $p (keys %$packages){
					eval { import $p };
					# die "$err is a method in $p"
					undef $err if $p->can( $err );
				}
			};
		}
		
		@err = grep {defined} @err;
		
		if (@err){
			push @rv, @err;
			$self->{errors}->[$line] = \@err;
		}
		$line ++;
	}

	$self->_remove_temporary_stoplist;

	return @rv;
}

sub add_allow_words {
	my $self = shift;
	push @{ $self->{allow_words} }, @_ if $#_ > -1;
}

sub skip_paths_matching {
	my $self = shift;
	push @{ $self->{skip_paths_matching} }, @_ if $#_ > -1;
	return @{ $self->{skip_paths_matching} };
}

sub _add_temporary_stoplist {
	my ($self, @stoplist) = @_;
	my $dict = { map {$_=>1} @{$self->{allow_words}} };
	my @new;
	foreach my $word (@stoplist){
		push @new, $word if not exists $dict->{$word};
	}
	push @{ $self->{_temp_stoplist} }, \@new;
	$self->add_allow_words( @new );
}

sub _remove_temporary_stoplist {
	my ($self, @stoplist) = @_;
	return if not scalar @{ $self->{_temp_stoplist} };
	my $remove = { map {$_=>1} pop @{ $self->{_temp_stoplist} } };
	my @allowed;
	foreach my $word (@{ $self->{allow_words} }){
		push @allowed, $word if not exists $remove->{$word};
	}
}

1;

=encoding utf8

=head1 NAME

Pod::Spelling - Send POD to a spelling checker

=head1 SYNOPSIS

	use Pod::Spelling;
	my $o = Pod::Spelling->new();
	say 'Spelling errors: ', join ', ', $o->check_file( 'Module.pm' );

	use Pod::Spelling;
	my $o = Pod::Spelling->new( import => 'My::Speller' );
	say 'Spelling errors: ', join ', ', $o->check_file( 'Module.pm' );

	use Pod::Spelling;
	my $o = Pod::Spelling->new(
		allow_words => [qw[ foo bar ]],
	);
	$o->skip_paths_matching( qr{*/DBIC} );
	say 'Spelling errors: ', join ', ', $o->check_file( 'Module.pm' );

=head1 DESCRIPTION

This module provides extensible spell-checking of POD.

At present, it checks for (in order) one of L<Text::Hunspell>, 
L<Lingua::Ispell> or L<Text::Aspell>,
one of which must be installed on your system, with its binaries, 
unless you plan to use the API to provide your own spell-checker. In
the latter case, or if binaries are missing from their default locations,
expect test failures.

=head1 TEXT NOT SPELL-CHECKED

The items below commonly upset spell-checking, though are generally
considered valid in POD, and so are not sent to the spell-checker.

=over 4

=item *

The body of links (C<LE<lt>...E<gt>>) and file-formatted strings (C<FE<lt>...E<gt>>).

=item *

Verbatim blocks (indented text, as used in C<SYNOPSIS> sections.

=item *

Any string containing two colons (C<::>).

=item *

The name of the module as written in the standard POD manner:

	=head1 NAME
	
	Module::Name::Here - brief description here
	
=item *

Words contained in L<Pod::Wordlist|Pod::Wordlist>, though that can be disabled
- see the C<no_pod_wordlist>, below.

=back
	
=head1 CONSTRUCTOR (new)

Optional parameters:

=over 4

=item C<allow_words>

A list of words to remove from text prior to it being spell-checked.

=item C<no_pod_wordlist>

Prevents the default behaviour of using L<Pod::Wordlist|Pod::Wordlist> 
to ignore words often used in Perl modules, but rarely found in dictionaries.

=item C<import_speller>

Name of a class to that implements
the C<_init> method and the C<Pod::Spelling::_spell_check_callback> method.
Current implementations are L<Pod::Spelling::Ispell|Pod::Spelling::Ispell>
and L<Pod::Spelling::Aspell|Pod::Spelling::Aspell>. If anything else should be
added, please let me know.

=back

If no C<import_speller> is specified, then C<Ispell> is tried, then C<Aspell>,
then the module croaks.

=head1 DEPENDENCIES

L<Pod::POM|Pod::POM>.

=head1 METHODS

=head2 check_file

Accepts a path to a file, runs the spell check, and returns a list of badly-spelt
words, setting the C<errors> field with an array, each entry of which is a list that
represents a line in the file, and thus may be empty if there are no spelling errors.

=head2 add_allow_words

Add a list of words to the 'allow' list specified at instantiation.

=head2 skip_paths_matching

Supply a list of one or more pre-compiled regular expressions to
avoid parsing directories they match.

=head1 ADDING A SPELL-CHECKER

This module is really just a factory class that does nothing but 
provide an API for sending POD to a spelling checker via a callback method,
and returning the results. 

The spell-checking callback method, supplied as a
code reference in the C<spell_check_callback> argument during construction,
receives a list of text, and should return a list of badly-spelt words.

	my $o = Pod::Spelling->new(
		spell_check_callback => sub { 
			my ($self, @text) = @_;
			return $find_bad_words( \@text );
		},
	);

Alternatively, this module can be sub-classed: see the source of
C<Pod::Spelling::Ispell>.

=head1 SEE ALSO

L<Test::Pod::Spelling>,
L<Pod::Spelling::Hunspell>,
L<Pod::Spelling::Aspell>,
L<Pod::Spelling::Ispell>,
L<Pod::POM>,
L<Pod::POM::View::TextBasic>,
L<Pod::Spell>,
L<Pod::WordList>.

=head1 AUTHOR AND COPYRIGHT

Copyright (C) Lee Goddard, 2011. All Rights Reserved.

Made available under the same terms as Perl.




lib/Pod/Spelling/Aspell.pm
use strict;
use warnings;

package Pod::Spelling::Aspell;
use base 'Pod::Spelling';
use Carp;
no warnings 'redefine';

our $VERSION = 0.1;

require Text::Aspell;

sub _init {
	shift if not ref $_[0];
	my $self = shift;
	
	$self->{aspell} ||= Text::Aspell->new
		or die 'Could not instantiate Text::Aspell';
		
	$self->{aspell}->check('house');
	return $self->{aspell}->errstr if $self->{aspell}->errstr;

	return $self;
}

# Accepts one or more lines of text, returns a list mispelt words.
sub _spell_check_callback {
	my ($self, @lines) = @_;
	my $errors;
	for my $word ( split /\s+/, join( ' ', @lines ) ){
		next if not $word;
		$errors->{ $word } ++ if not $self->{aspell}->check( $word )
			and not $self->{aspell}->errstr;
	}
	
	return keys %$errors;
}

1;

__END__

=head1 NAME

Pod::Spelling::Aspell - Spell-test POD with Aspell

=head1 SYNOPSIS

	my $o = Pod::Spelling::Apsell->new(
		allow_words => qw[ Django Rheinhardt ],
	);
	warn "Spelling errors: ", join ', ', $o->check_file( 'blib/Paris.pm' );

=head1 DESCRIPTION

Checks the spelling in POD using the C<aspell> program, which is expected
to be found on the system. 

You may configure and supply an instance of C<Text::Aspell> to the constructor
with the C<aspell> argument.

For details of options and methods, see the parent class,
L<Pod::Spelling|Pod::Spelling>.

=head1 SEE ALSO

L<Pod::Spelling>, L<Test::Pod::Spelling>.

=head1 AUTHOR

Lee Goddard (C<lgoddard-at-cpan.org>)

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2011, Lee Goddard. All Rights Reserved.

Made available under the same terms as Perl.





lib/Pod/Spelling/Ispell.pm
use strict;
use warnings;

package Pod::Spelling::Ispell;

use base 'Pod::Spelling';
no warnings 'redefine';

our $VERSION = 0.2;

require Lingua::Ispell;

Lingua::Ispell::allow_compounds(1);

sub _init {
	shift if not ref $_[0];
	my $self = shift;
	if (not defined $self->{ispell_path}){
		# Which ispell?
		foreach my $p (
			'', # Already on the PATH?
			qw( 
			/usr/local/bin/
			/usr/local/sbin/
			/usr/bin/
			/opt/usr/bin/
			/opt/local/bin/
		)){
			if (-e $p.'ispell'){
				no warnings;
				$self->{ispell_path} = $p.'ispell';
			}
		}
	}

	if (defined $self->{ispell_path}){
		$Lingua::Ispell::path = $self->{ispell_path} 
	} else {
		$self->{spell_check_callback} = undef;	
	}

	return $self;
}


# Accepts one or more lines of text, returns a list mispelt words.
sub _spell_check_callback {
	my $self = shift;
	my @lines = @_;
	my $errors;
	
	for my $r ( Lingua::Ispell::spellcheck( @lines )){
		$errors->{ $r->{term} } ++ if $r->{type} =~ /^(miss|guess|none)$/;
	}
	
	return keys %$errors;
}

1;

__END__

=head1 NAME

Pod::Spelling::Ispell - Spell-test POD with Ispell

=head1 SYNOPSIS

	my $o = Pod::Spelling::Ispell->new(
		allow_words => qw[ Django Rheinhardt ],
	);
	warn "Spelling errors: ", join ', ', $o->check_file( 'blib/Paris.pm' );

=head1 DESCRIPTION

Checks the spelling in POD using the C<ispell> program, which is expected
to be found on the system. 

When calling the constructor, you may supply the argument C<ispell_path>
to specify the full path to the C<ispell> executable. This module has a guess,
but may have missed possible locations.

For details of options and methods, see the parent class,
L<Pod::Spelling|Pod::Spelling>.

=head1 SEE ALSO

L<Pod::Spelling>, L<Test::Pod::Spelling>, L<Lingua::Ispell>.

=head1 AUTHOR

Lee Goddard (C<lgoddard-at-cpan.org>)

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2011, Lee Goddard. All Rights Reserved.

Made available under the same terms as Perl.





lib/Pod/Spelling/Hunspell.pm
use strict;
use warnings;

package Pod::Spelling::Hunspell;

use base 'Pod::Spelling';
no warnings 'redefine';

use Carp;
require Text::Hunspell;

our $VERSION = 0.1;


# The default dictionary depends on the locale settings. The following
# environment variables are searched: 
#
#		LC_ALL, LC_MESSAGES, and LANG. 
#
# If none are set then the following fallbacks are used:
#
# /usr/share/myspell/default.aff Path of default affix file. See hunspell(4).
# /usr/share/myspell/default.dic Path of default dictionary file. See hunspell(4).
# $HOME/.hunspell_default. Default path to personal dictionary.

our $CONFIG_ROOT_PATH;

sub _guess_paths {
	my @langs;
	my $CONFIG_ROOT_PATH;

	my @paths = qw(
		/usr/share/hunspell/
		/usr/local/bin/myspell
		/usr/local/sbin/myspell
		/usr/bin/myspell
		/opt/usr/bin/myspell
		/opt/local/bin/myspell
	);


	foreach my $i (qw( LC_ALL LC_MESSAGES LANG )){
		next unless $ENV{$i};
		push @langs, $ENV{$i};
		my ($gen) = $ENV{$i} =~ /([^.]+)\.[^.]+$/;
		push @langs, $gen if $gen;
	}

	@langs = (qw( en_GB en_US )) if not @langs;

	my @full_paths;

	foreach my $path (@paths){
		foreach my $lang (@langs){
			push @full_paths, $path .'/'.$lang;
		}
	}

	push @full_paths, 
		'/usr/share/myspell/default',
		'/usr/share/hunspell/default',
		($ENV{HOME}? $ENV{HOME}.'/.hunspell_default' : ()),
		($ENV{HOME}? $ENV{HOME}.'/.myspell_default' : ());

	foreach my $path (@full_paths){
		next if not -e $path .'.dic';
		next if not -e $path .'.aff';
		$CONFIG_ROOT_PATH = $path;
		last;
	}

	return $CONFIG_ROOT_PATH;
}

sub _init {
	shift if not ref $_[0];
	my $self = shift || {};

	if (not defined $self->{hunspell_aff} and not defined $self->{hunspell_dic} ){
		$CONFIG_ROOT_PATH = _guess_paths() if not $CONFIG_ROOT_PATH;
		confess 'No hunspell_aff and hunspell_dic, and could not even guess at paths' if not $CONFIG_ROOT_PATH;
		if (-e $CONFIG_ROOT_PATH .'.aff' and -e $CONFIG_ROOT_PATH .'.dic'){
			$self->{hunspell_aff} = $CONFIG_ROOT_PATH.'.aff';
			$self->{hunspell_dic} = $CONFIG_ROOT_PATH.'.dic';
		} 
		else {
			confess 'Guessed path root was useless, ', $CONFIG_ROOT_PATH, '*';
		}

		$self->{hunspell} = Text::Hunspell->new( $self->{hunspell_aff}, $self->{hunspell_dic} );
		if (not $self->{hunspell}->check('house')){
			confess "Hunspell could not spell 'house',\n",
				"\tusing aff ",$self->{hunspell_aff},"\n",
				"\tusing dic ",$self->{hunspell_dic};
		} 
	}

	# if (not defined $self->{hunspell_aff} or not defined $self->{hunspell_dic}){
	# 	$self->{spell_check_callback} = undef;	
	# }

	return $self;
}


# Accepts one or more lines of text, returns a list mispelt words.
sub _spell_check_callback {
	my ($self, @lines) = @_;
	my $errors;

	for my $word ( split /\s+/, join( ' ', @lines ) ){
		next if not $word;
		next if length $word < 2;

		if (not $self->{hunspell}->check($word)){
			$errors->{ $word } ++;
		}
	}
	
	return keys %$errors;
}

1;

__END__

=head1 NAME

Pod::Spelling::Hunspell - Spell-test POD with hunspell

=head1 SYNOPSIS

	my $o = Pod::Spelling::Hunspell->new(
		allow_words => qw[ Django Rheinhardt ],
	);
	warn "Spelling errors: ", join ', ', $o->check_file( 'blib/Paris.pm' );

=head1 DESCRIPTION

Checks the spelling in POD using the C<hunspell> program, which is expected
to be found on the system. 

When calling the constructor, you may supply the arguments C<hunspell_dic>
and C<hunspell_aff> to specify the full paths to the C<hunspell> dictionaries. 
Otherwise, this module will have a guess, but may have missed possible locations.

For details of options and methods, see the parent class,
L<Pod::Spelling|Pod::Spelling>.

=head1 SEE ALSO

L<Pod::Spelling>, L<Test::Pod::Spelling>.

=head1 AUTHOR

Lee Goddard (C<lgoddard-at-cpan.org>)

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2011, Lee Goddard. All Rights Reserved.

Made available under the same terms as Perl.





lib/Pod/POM/View/TextBasic.pm
use strict;
use warnings;

package Pod::POM::View::TextBasic;
our $VERSION = 0.2;

use base 'Pod::POM::View::Text';

our $DROPS = 1;

sub view_for { 
    my ($format, $text) = $_[1] =~ /^\s*\=for\s+(\S+)\s+(.*?)\s*$/si;
    $text =~ s/\s+/ /g;
    return $format =~ /^(stopwords?|spelling|allow|allow_words|allowed_words)$/gi? $text : '';
}

sub view_seq_bold { return $_[1]; }

sub view_seq_italic { return $_[1]; }

sub view_seq_code { return $DROPS? '' : $_[1] }

sub view_seq_file { return $DROPS? '' : $_[1] }

sub view_seq_link {
	return $DROPS? '' : $_[1];
    my ($self, $link) = @_;
    return ($link =~ m/^(.*?)\|/) ?
    	$1 : $link;
}
	
sub view_verbatim { return $DROPS? '' : $_[1] }

1;

=head1 NAME

Pod::POM::View::TextBasic - Pod::POM::View::Text without sequence formatting

=head1 SYNOPSIS

	Pod::POM->default_view( 'Pod::POM::View::TextBasic' )
		or die $Pod::POM::ERROR;
	my $p = Pod::POM->new;
	$p->parse_file(...);
	
=head1 DESCRIPTION

A sub-class of L<Pod::POM::View::Text|Pod::POM::View::Text> 
that does not ornament inline sequences to indicate their presence.

This class was developed to aid spell-checking POD: for that reason,
verbatim blocks, inline code sequences, and the bodies of links are ignored 
unless C<Pod::POM::View::TextBasic::DROPS> is set to a false value by the user.

=head1 AUTHOR

Lee Goddard (C<lgoddard-at-cpan.org>).

=head1 COPYRIGHT 

Copyright (C) 2011, Lee Goddard. All Rights Reserved.





Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker;
use Data::Dumper;

my $ok;
my @errors;

use lib qw( lib );

{
	eval "require Text::Hunspell";
	if ($@){
		push @errors, 'No Text::Hunspell';
	} 
	
	else {
		my $path;
		FIND: 
		$ok = undef;

		require Pod::Spelling::Hunspell;
		eval { Pod::Spelling::Hunspell->_init };
		if ($@){
			push @errors, $@;
		} else {
			$ok = 1;
		}
	}
}

if (not $ok){
	eval "require Text::Aspell";
	if ($@){
		push(@errors, "No Text::Aspell installed");
		last;
	}
	
	my $aspell = Text::Aspell->new;
	if (not $aspell){
		push @errors, 'Could not make Aspell object';
		last;
	}

	$aspell->check('house');
	if ($aspell->errstr){
		push @errors, 'Aspell: ' . $aspell->errstr;
		last;
	}
	
	$ok = 1;
}

if (not $ok){
	eval "require Lingua::Ispell";
	if ($@){
		push @errors, 'No Ispell';
	} 
	
	elsif (Lingua::Ispell->VERSION < 0.07){
		push @errors, 'Sorry, only Linuga::Ispell version >= 0.07 is supported ';
	}
	
	else {
		
		FIND: 
		foreach my $p (
			'', # Already on the PATH?
			qw( 
				/usr/local/bin/
				/usr/local/sbin/
				/usr/bin/
				/opt/usr/bin/
				/opt/local/bin/
			)
		){
			my $path = $p . ($^O eq 'MSWin32'? 'ispell.exe' : 'ispell');
			if (`$path -v`){
				no warnings;
				$Lingua::Ispell::path = $path;
				$ok = 1;
				last FIND;
			}
			push @errors, "Could not find ispell as $path";
		}
		
		if ($ok){
			$ok = undef;
			my $r;
			eval { $r = Lingua::Ispell::spellcheck( 'house' ) };
			if ($@){
				push @errors, "Ispell died: $@";
			}
			elsif (ref $r) { # API changed, no rv for correect spelling: $r->{type} =~ /^(miss|guess|none)$/){
				push @errors, "Ispell could not spell 'house'";
			} 
			else {
				$ok = 1;
			}
		}
	}
}


if (not $ok){
	print <<_STOP_;

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

This module  requires you  have  installed  on your system
one of the following Perl modules and their binaries:

	Text::Hunspell
    Lingua::Ispell
    Text::Aspell

If you feel this  message to be  in  error, please contact
the author via LGoddard-at-CPAN-dot-org.

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

_STOP_

	warn join "\n", @errors;
	warn "Writing Makefile anyway\n";
} 

# else 
{
	
	WriteMakefile(
	  # ( MM->can( 'signature_target' ) ? ( SIGN => 1 ) : () ),
	  license( 'perl' ),
	  NAME          => 'Pod::Spelling',
	  AUTHOR        => 'Lee Goddard <lgoddard@cpan.org>',
	  VERSION		=> '0.23',
	  ABSTRACT_FROM => 'lib/Pod/Spelling.pm',
	  PL_FILES      => {},
	  PREREQ_PM     => {
		'Pod::POM'	=> 0.04,
		'Test::Pod' => 0,
		'Test::Builder' => 0,
	  },
	  dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
	  clean => { FILES    => 'Pod-Spelling-*' },
	);

}

sub license {
  my $lic = shift;
  local $^W = 0;    # Silence warning about non-numeric version
  return
   unless $ExtUtils::MakeMaker::VERSION >= '6.31';
  return ( LICENSE => $lic );
}



